<a href="https://colab.research.google.com/github/pritiyadav888/Machine-Learning-Projects/blob/main/Broadcast_in_python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

X is a 3*4 matrix.

In [1]:
import numpy as np
import time
X = np.array([[2, 4, 6, 8],[3.45, 2.39, 9.1, 7.7],[33, 54, 76, 90]])
print(X)

[[ 2.    4.    6.    8.  ]
 [ 3.45  2.39  9.1   7.7 ]
 [33.   54.   76.   90.  ]]


Let's turn our matrix into a dataframe with row names and columns for explaination purposes

In [2]:
import pandas as pd
column_names= ["test1", "test2", "test3", "test4"]
row_names = ["maths", "physics", "chemistry"]
df = pd.DataFrame(X, columns=column_names, index=row_names)
print(df)

           test1  test2  test3  test4
maths       2.00   4.00    6.0    8.0
physics     3.45   2.39    9.1    7.7
chemistry  33.00  54.00   76.0   90.0


Calculate percentage of maths, physics and chemistry marks for each test respectively out of 100 total marks. Do it without a loop. Find out why axis is equal to 0. 

In [3]:
cal = X.sum(axis=0)
print(cal)

[ 38.45  60.39  91.1  105.7 ]


Here comes the broadcast

In [5]:
percent = 100*df/cal.reshape(1,4)
print(percent)

               test1      test2      test3      test4
maths       5.201560   6.623613   6.586169   7.568590
physics     8.972692   3.957609   9.989023   7.284768
chemistry  85.825748  89.418778  83.424808  85.146641
18.97287368774414


Yay! You have a solution in the cell above. In test1 - percentage of maths is approx 5.2, physics is approx 8.9 and chemistry is 85.8. We did it without an explicit for loop. As we want python to the sum vertically, so axis is 0. Check out the for loop for this execise below.

In [7]:
 for j in range(len(X[0])):
    for i in range(len(X)):
      print( X[i][j]/cal[j]*100 )

5.201560468140442
8.972691807542263
85.8257477243173
6.6236131809902306
3.9576088756416627
89.41877794336811
6.586169045005488
9.989023051591658
83.42480790340286
7.5685903500473035
7.28476821192053
85.14664143803216
